tests/e2e_tests.c

```diff
@@ -49,48 +53,82 @@ static void	finally(int *out, int *save)
 	close(*save);
 }
 
+#ifdef FT_CALLOC_IF_TRAPPED
 static void	finally_err(int *out, int *save)
 {
 	fflush(stderr);
 	close(*out);
 	dup2(*save, fileno(stderr));
 	close(*save);
 }
+#endif
 
 static int	file_compare(char *exp_contens, char *act_fname)
 {
-	int fc = open(act_fname, O_RDONLY, 0600);
-	int len_exp = ft_strlen(exp_contens);
-	char *act = ft_calloc(sizeof(char), len_exp * 2 + 1);
+	#ifdef DEBUG
+	fprintf(stderr, "comparing files\n");
+	fflush(stderr);
+	fprintf(stderr, "expected [%p]\n", exp_contens);
+	fflush(stderr);
+	fprintf(stderr, "expected [%s]\n", exp_contens);
+	fflush(stderr);
+	fprintf(stderr, "actual fname [%s]\n", act_fname);
+	fflush(stderr);
+	#endif
+	FILE *f = fopen(act_fname, "rb");
+	fseek(f, 0, SEEK_END);
+	long fsize = ftell(f);
+	fseek(f, 0, SEEK_SET);
+	char *act = ft_calloc(sizeof(char), fsize + 1);
 	assert(act);
-	int len_act = read(fc, act, len_exp * 2);
-	assert(len_act - len_exp <= 3); // expected result doesn't store a number
-	int comp_res = ft_strncmp(exp_contens, act, ft_strlen(exp_contens) - 1);
+	fread(act, fsize, 1, f);
+	act[fsize] = 0;
+	fclose(f);
+
+	regex_t re;
+	char	*exp_re = ft_strjoin(exp_contens, "(Total ft_calloc_calls: [0-9]*\n)?");
+	assert(exp_re);
+	assert(regcomp(&re, exp_re, REG_EXTENDED | REG_NOSUB) == 0);
+	int comp_res = regexec(&re, act, 0, NULL, 0);
 	#ifdef DEBUG
-	ft_printf("comparison result %i, expected:[%s] (%i symbols)\nactual:[%s] (%i symbols)\n", comp_res, exp_contens, len_exp, act, len_act);
+	int	diff = -1;
+	for (int i = 0; i < fsize; i ++)
+		if (act[i] != exp_re[i])
+		{
+			diff = i;
+			break ;
+		}
+	ft_printf("comparison result %i, expected:{{%s}}\nactual:{{%s}}\nstrncmp %i\nstart to differ from index %i\n", comp_res, exp_re, act, ft_strncmp(exp_contens, act, fsize), diff);
+	fflush(stdout);
 	#endif
 	assert(comp_res == 0);
+	free(exp_re);
 	int i;
-	for (i = len_act - 2; act[i] >= '0' && act[i] <= '9'; i --);
-	char *sub = ft_substr(act, i, len_act);
+	for (i = fsize - 2; act[i] >= '0' && act[i] <= '9'; i --);
```

Comment:
the role of it please

Reply:
When started with traps, the output always ends in a number. Counter of callocs. Here I navigate fron the end of file to the last non-digit to find the start of this number to parse it out.